File: Front/lib/core/models/goal.dart
/// Tipos de metas financeiras
enum GoalType {
  savings('SAVINGS', 'Economizar', '💰'),
  expenseReduction('EXPENSE_REDUCTION', 'Reduzir Gastos', '📉'),
  incomeIncrease('INCOME_INCREASE', 'Aumentar Receita', '📈'),
  custom('CUSTOM', 'Personalizada', '✏️');

  const GoalType(this.value, this.label, this.icon);
  final String value;
  final String label;
  final String icon;
}

/// Modelo de Meta Financeira
class GoalModel {
  const GoalModel({
    required this.id,
    required this.title,
    required this.description,
    required this.targetAmount,
    required this.currentAmount,
    this.initialAmount = 0.0,
    this.deadline,
    required this.goalType,
    required this.progressPercentage,
    required this.createdAt,
    required this.updatedAt,
    this.targetCategories = const [],
    this.targetCategoryNames = const [],
    this.baselineAmount,
    this.trackingPeriodMonths = 3,
  });

  final String id;
  final String title;
  final String description;
  final double targetAmount;
  final double currentAmount;
  final double initialAmount;
  final DateTime? deadline;
  final GoalType goalType;
  final double progressPercentage;
  final DateTime createdAt;
  final DateTime updatedAt;
  
  // Suporte a múltiplas categorias
  final List<String> targetCategories;     // Lista de IDs de categorias
  final List<String> targetCategoryNames;  // Lista de nomes (read-only)
  final double? baselineAmount;            // Valor de referência inicial
  final int trackingPeriodMonths;          // Período de cálculo em meses

  /// Compatibilidade retroativa: primeira categoria
  String? get targetCategory => targetCategories.isNotEmpty ? targetCategories.first : null;
  
  /// Compatibilidade retroativa: nome da primeira categoria
  String? get targetCategoryName => targetCategoryNames.isNotEmpty ? targetCategoryNames.first : null;

  factory GoalModel.fromMap(Map<String, dynamic> map) {
    // Parse target_categories (lista de IDs)
    List<String> categories = [];
    if (map['target_categories'] != null) {
      categories = (map['target_categories'] as List)
          .map((e) => e.toString())
          .toList();
    } else if (map['target_category'] != null) {
      // Compatibilidade retroativa
      categories = [map['target_category'].toString()];
    }
    
    // Parse target_category_names (lista de nomes)
    List<String> categoryNames = [];
    if (map['target_category_names'] != null) {
      categoryNames = (map['target_category_names'] as List)
          .map((e) => e.toString())
          .toList();
    } else if (map['target_category_name'] != null) {
      // Compatibilidade retroativa
      categoryNames = [map['target_category_name'].toString()];
    }
    
    return GoalModel(
      id: map['id'].toString(),
      title: (map['title'] as String?) ?? '',
      description: (map['description'] as String?) ?? '',
      targetAmount: double.tryParse(map['target_amount']?.toString() ?? '0') ?? 0.0,
      currentAmount: double.tryParse(map['current_amount']?.toString() ?? '0') ?? 0.0,
      initialAmount: map['initial_amount'] != null
          ? double.tryParse(map['initial_amount'].toString()) ?? 0.0
          : 0.0,
      deadline: map['deadline'] != null
          ? DateTime.tryParse(map['deadline'] as String)
          : null,
      goalType: _parseGoalType(map['goal_type'] as String?),
      progressPercentage:
          double.tryParse(map['progress_percentage']?.toString() ?? '0') ?? 0.0,
      createdAt: map['created_at'] != null 
          ? DateTime.tryParse(map['created_at'] as String) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: map['updated_at'] != null
          ? DateTime.tryParse(map['updated_at'] as String) ?? DateTime.now()
          : DateTime.now(),
      // Múltiplas categorias
      targetCategories: categories,
      targetCategoryNames: categoryNames,
      baselineAmount: map['baseline_amount'] != null
          ? double.tryParse(map['baseline_amount'].toString())
          : null,
      trackingPeriodMonths: map['tracking_period_months'] as int? ?? 3,
    );
  }

  static GoalType _parseGoalType(String? value) {
    switch (value?.toUpperCase()) {
      case 'SAVINGS':
      case 'EMERGENCY_FUND':  // Compatibilidade: tratar como SAVINGS
        return GoalType.savings;
      case 'EXPENSE_REDUCTION':
        return GoalType.expenseReduction;
      case 'INCOME_INCREASE':
        return GoalType.incomeIncrease;
      default:
        return GoalType.custom;
    }
  }

  Map<String, dynamic> toMap() {
    return {
      'title': title,
      'description': description,
      'goal_type': goalType.value,
      'target_amount': targetAmount.toString(),
      if (initialAmount > 0) 'initial_amount': initialAmount.toString(),
      if (deadline != null) 
        'deadline': deadline!.toIso8601String().split('T')[0],
      // Múltiplas categorias
      if (targetCategories.isNotEmpty) 'target_categories': targetCategories,
      if (baselineAmount != null) 'baseline_amount': baselineAmount.toString(),
      'tracking_period_months': trackingPeriodMonths,
    };
  }

  double get progress =>
      targetAmount == 0 ? 0 : (currentAmount / targetAmount).clamp(0, 1);

  /// Retorna o identificador preferencial
  String get identifier => id;
  
  /// Verifica se possui UUID
  bool get hasUuid => true;

  /// Verifica se a meta está completa
  bool get isCompleted => progressPercentage >= 100;

  /// Verifica se o prazo expirou
  bool get isExpired {
    if (deadline == null) return false;
    return DateTime.now().isAfter(deadline!);
  }

  /// Dias restantes até o prazo
  int? get daysRemaining {
    if (deadline == null) return null;
    final diff = deadline!.difference(DateTime.now());
    return diff.inDays;
  }

  /// Valor restante para atingir a meta
  double get amountRemaining => (targetAmount - currentAmount).clamp(0, double.infinity);
  
  /// Nome(s) das categorias formatado para exibição
  String get categoriesDisplayName {
    if (targetCategoryNames.isEmpty) return '';
    if (targetCategoryNames.length == 1) return targetCategoryNames.first;
    return '${targetCategoryNames.first} +${targetCategoryNames.length - 1}';
  }
}

File: Api/finance/models/goal.py
"""
Modelo Goal - Metas financeiras do usuário.
"""

from datetime import timedelta
from decimal import Decimal
import uuid

from django.conf import settings
from django.db import models
from django.utils import timezone

from .base import (
    MAX_AMOUNT,
    MAX_TITLE_LENGTH,
    MAX_DEADLINE_YEARS,
)


class Goal(models.Model):
    """
    Meta financeira do usuário.
    
    Tipos de metas:
    - SAVINGS: Juntar dinheiro (acumular valor)
    - EXPENSE_REDUCTION: Reduzir gastos em uma categoria
    - INCOME_INCREASE: Aumentar receita
    - EMERGENCY_FUND: Fundo de emergência (reserva de X meses)
    - CUSTOM: Meta personalizada (atualização manual)
    """
    
    class GoalType(models.TextChoices):
        SAVINGS = "SAVINGS", "Juntar Dinheiro"
        EXPENSE_REDUCTION = "EXPENSE_REDUCTION", "Reduzir Gastos"
        INCOME_INCREASE = "INCOME_INCREASE", "Aumentar Receita"
        EMERGENCY_FUND = "EMERGENCY_FUND", "Fundo de Emergência"
        CUSTOM = "CUSTOM", "Personalizada"

    id = models.UUIDField(
        primary_key=True,
        default=uuid.uuid4,
        editable=False,
        help_text="Identificador único universal (UUID v4)"
    )
    
    user = models.ForeignKey(
        settings.AUTH_USER_MODEL, 
        on_delete=models.CASCADE, 
        related_name="goals"
    )
    title = models.CharField(max_length=MAX_TITLE_LENGTH)
    description = models.TextField(blank=True)
    target_amount = models.DecimalField(max_digits=12, decimal_places=2)
    current_amount = models.DecimalField(
        max_digits=12, 
        decimal_places=2, 
        default=Decimal("0.00")
    )
    initial_amount = models.DecimalField(
        max_digits=12,
        decimal_places=2,
        default=Decimal("0.00"),
        help_text="Valor inicial da meta (transações anteriores à criação)"
    )
    deadline = models.DateField(null=True, blank=True)
    
    goal_type = models.CharField(
        max_length=20,
        choices=GoalType.choices,
        default=GoalType.CUSTOM,
        help_text="Tipo da meta"
    )
    
    target_categories = models.ManyToManyField(
        'Category',
        blank=True,
        related_name='goals_targeting_this',
        help_text="Categorias alvo para metas de redução de gastos (máximo 5)"
    )
    
    @property
    def target_category(self):
        """Compatibilidade retroativa: retorna a primeira categoria."""
        return self.target_categories.first()
    
    @property
    def target_category_name(self):
        """Nome da primeira categoria (compatibilidade)."""
        cat = self.target_categories.first()
        return cat.name if cat else None
    
    baseline_amount = models.DecimalField(
        max_digits=12,
        decimal_places=2,
        null=True,
        blank=True,
        help_text="Valor de referência inicial (gasto/receita mensal antes da meta)"
    )
    
    tracking_period_months = models.PositiveIntegerField(
        default=3,
        help_text="Período em meses para cálculo de progresso (padrão: 3 meses)"
    )
    
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        ordering = ("-created_at",)
        verbose_name = "Meta"
        verbose_name_plural = "Metas"

    def __str__(self) -> str:
        return f"{self.title} ({self.user})"

    def clean(self):
        from django.core.exceptions import ValidationError
        
        self._validate_amounts()
        self._validate_title()
        self._validate_deadline()
        self._validate_goal_type_fields()

    def _validate_amounts(self):
        from django.core.exceptions import ValidationError
        
        if self.target_amount is not None and self.target_amount <= 0:
            raise ValidationError({
                'target_amount': 'O valor alvo da meta deve ser maior que zero.'
            })
        
        if self.target_amount is not None and self.target_amount > MAX_AMOUNT:
            raise ValidationError({
                'target_amount': f'O valor alvo excede o limite máximo (R$ {MAX_AMOUNT:,.2f}).'
            })
        
        if self.current_amount is not None and self.current_amount < 0:
            raise ValidationError({
                'current_amount': 'O valor atual não pode ser negativo.'
            })
        
        if self.initial_amount is not None and self.initial_amount < 0:
            raise ValidationError({
                'initial_amount': 'O valor inicial não pode ser negativo.'
            })

    def _validate_title(self):
        from django.core.exceptions import ValidationError
        
        if self.title and len(self.title.strip()) == 0:
            raise ValidationError({
                'title': 'O título não pode ser vazio.'
            })
        
        if self.title and len(self.title) > MAX_TITLE_LENGTH:
            raise ValidationError({
                'title': f'O título não pode ter mais de {MAX_TITLE_LENGTH} caracteres.'
            })

    def _validate_deadline(self):
        from django.core.exceptions import ValidationError
        
        if self.deadline:
            if self.deadline < timezone.now().date():
                raise ValidationError({
                    'deadline': 'A data limite deve ser no futuro.'
                })
            max_deadline = timezone.now().date() + timedelta(days=MAX_DEADLINE_YEARS * 365)
            if self.deadline > max_deadline:
                raise ValidationError({
                    'deadline': f'A data limite não pode ser mais de {MAX_DEADLINE_YEARS} anos no futuro.'
                })
    
    def _validate_goal_type_fields(self):
        """Valida campos específicos por tipo de meta."""
        from django.core.exceptions import ValidationError
        
        if self.goal_type == Goal.GoalType.EXPENSE_REDUCTION:
            # Nota: validação de target_categories é feita no serializer
            # pois ManyToMany não está disponível antes do save()
            
            if not self.baseline_amount or self.baseline_amount <= 0:
                raise ValidationError({
                    'baseline_amount': 'Informe o valor médio mensal atual de gastos nesta categoria.'
                })
            
            # Validar que target_amount faz sentido em relação ao baseline
            if self.baseline_amount and self.target_amount:
                if self.target_amount >= self.baseline_amount:
                    raise ValidationError({
                        'target_amount': 'A meta de redução deve ser menor que o valor base atual.'
                    })
        
        elif self.goal_type == Goal.GoalType.INCOME_INCREASE:
            if not self.baseline_amount or self.baseline_amount <= 0:
                raise ValidationError({
                    'baseline_amount': 'Informe sua receita média mensal atual para comparação.'
                })
        
        # SAVINGS, EMERGENCY_FUND e CUSTOM não precisam de validações extras


    @property
    def progress_percentage(self) -> float:
        if self.target_amount == 0:
            return 0.0
        return min(100.0, (float(self.current_amount) / float(self.target_amount)) * 100.0)
